### PR TITLE
Internal: Editor V2 - Route listeners [ED-9435]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useState } from 'react';
+import { listenTo, routeCloseEvent, routeOpenEvent, RouteEventDescriptor } from '../listeners';
+
+export function useIsRouteActive( route: RouteEventDescriptor['name'] ): boolean {
+	const [ isActive, setIsActive ] = useState( isRouteActive( route ) );
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:5]:
```suggestion
	const [ isActive, setIsActive ] = useState( () => isRouteActive( route ) );
```

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+	useEffect( () => {
+		const cleanupOpen = listenTo(
+			routeOpenEvent( route ),
+			() => setIsActive( true ),
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:10]:
```suggestion
			() => setIsActive( isRouteActive( route ) ),
```

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useState } from 'react';
+import { listenTo, routeCloseEvent, routeOpenEvent, RouteEventDescriptor } from '../listeners';
+
+export function useIsRouteActive( route: RouteEventDescriptor['name'] ): boolean {
+	const [ isActive, setIsActive ] = useState( isRouteActive( route ) );
+
+	useEffect( () => {
+		const cleanupOpen = listenTo(
+			routeOpenEvent( route ),
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:9]:
```suggestion
			[routeOpenEvent( route ), routeCloseEvent( route )],
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts (hunk 1 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:5]: Oops, done

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts (hunk 2 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:10]: Done

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts (hunk 3 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:9]: Thanks! much cleaner :D

File: packages/packages/v1-adapters/src/listeners/listeners.ts [hunk 4]
```diff
@@ -18,6 +18,7 @@ export function listenTo(
 		eventDescriptors = [ eventDescriptors ];
 	}
 
+	// eslint-disable-next-line array-callback-return -- Clashes with typescript.
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/listeners/listeners.ts:21]:
```suggestion
	// @see https://github.com/typescript-eslint/typescript-eslint/issues/2841
	// eslint-disable-next-line array-callback-return -- Clashes with typescript.
```

**Review by Nevoss — DISMISSED** [review]

**Review by Nevoss — DISMISSED** [review]

File: packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		act( () => {
+			mockedIsRouteActive.mockReturnValue( true );
+			dispatchRouteOpen( route );
+		} );
+
+		// Assert.
+		expect( result.current ).toBe( true );
+	} );
+
+	it( 'should return true when a route gets deactivated', () => {
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts:58]:
```suggestion
	it( 'should return false when a route gets deactivated', () => {
```

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import {
+	isRouteActive,
+	listenTo,
+	routeCloseEvent,
+	routeOpenEvent,
+	RouteEventDescriptor,
+} from '../';
+
+export function useIsRouteActive( route: RouteEventDescriptor['name'] ): boolean {
+	const [ isActive, setIsActive ] = useState( () => isRouteActive( route ) );
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:11]: The route status should be re-calculated when the route is changed.

File: packages/packages/v1-adapters/src/dispatchers/__tests__/index.test.ts [hunk 7]
```diff
@@ -102,6 +105,21 @@ describe( '@elementor/v1-adapters/dispatchers', () => {
...
+		const route = 'test/route';
+
+		( window as any ).$e.routes.isPartOf.mockReturnValue( true );
+
+		// Act.
+		const result = isRouteActive( route );
+
+		// Assert.
+		expect( result ).toEqual( true );
+		expect( ( window as any ).$e.routes.isPartOf ).toHaveBeenCalledTimes( 1 );
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/dispatchers/__tests__/index.test.ts:120]: Why do you need it?

File: packages/packages/v1-adapters/src/dispatchers/dispatchers.ts [hunk 8]
```diff
@@ -36,3 +36,13 @@ export function openRoute( route: string ): Promise<void> {
 		return Promise.reject( e );
 	}
 }
+
+export function isRouteActive( route: string ): boolean {
+	const extendedWindow = window as unknown as ExtendedWindow;
+
+	if ( ! extendedWindow.$e?.routes?.isPartOf ) {
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/dispatchers/dispatchers.ts:43]: Why do you need it? the returned check covers it.

File: packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+		const { result } = renderHook( () => useIsRouteActive( route ) );
+
+		act( () => {
+			mockedIsRouteActive.mockReturnValue( false );
+			dispatchRouteClose( route );
+		} );
+
+		// Assert.
+		expect( result.current ).toBe( false );
+	} );
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts:74]: Consider adding a test for a route change
+
A test for the hook cleanup?

**Review by rotemee — CHANGES_REQUESTED** [review]

File: packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts (hunk 5 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts:58]: Done

File: packages/packages/v1-adapters/src/dispatchers/__tests__/index.test.ts (hunk 7 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/dispatchers/__tests__/index.test.ts:120]: to make sure that I actually call the underlying $e.routes API properly

File: packages/packages/v1-adapters/src/hooks/use-is-route-active.ts (hunk 6 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/use-is-route-active.ts:11]: Done

File: packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts (hunk 9 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts:74]: Done

File: packages/packages/v1-adapters/src/dispatchers/dispatchers.ts (hunk 8 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/dispatchers/dispatchers.ts:43]: Done

File: packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+		// Assert.
+		expect( result.current ).toBe( true );
+
+		// Act.
+		rerender( {
+			route: 'inactive/route',
+		} );
+
+		rerender();
```
Comment (StyleShit) [inline packages/packages/v1-adapters/src/hooks/__tests__/use-is-route-active.test.ts:97]:
```suggestion
```

**Review by rotemee — APPROVED** [review]

**Pull request merged**